aiida_optimize/engines/_convergence.py

```diff
@@ -0,0 +1,236 @@
+        # to remove the calculation that is too rough from the window
+        # If we do, return -1 as an indication that convergence will not be
+        # possible
+        if self.current_index + num_new_iters > len(self.input_values):
+            # num_new_iters = len(self.input_values) - self.current_index
+            num_new_iters = -1
+
+        return num_new_iters
+
+    @property
+    def is_converged(self) -> bool:
+        """
+        Check if convergence has been reached by calculating the Frobenius
+        or 2-norm of the difference between all the result values / arrays
+        and checking that the maximum distance between points in the
+        convergence window is less than the tolerance.
+        """
+        if not self.initialized:
+            return False
+
+        # calculate pair distances between results
+        distance_triangle = self._distance_triangle
+        # flatten all the distances into a 1D list
+        distances = list(itertools.chain(*distance_triangle))
+
+        # check if the maximum distance is less than the tolerance
+        return np.max(distances) < self.tol
+
+    @property
+    def is_finished(self) -> bool:
+        if not self.initialized:
+            return False
+
+        # If we've used all the input values, we're finished
+        if len(self.result_values) >= len(self.input_values):
+            return True
+
+        # If _num_new_iters is -1, we know that we cannot converge with
+        # the remaining inputs in input_values, so we're finished
+        if self._num_new_iters == -1:
+            return True
+
+        return self.is_converged
+
+    @property
+    def is_finished_ok(self) -> bool:
+        if self.is_finished and self.is_converged:
+            return True
+        return False
+
+    def _create_inputs(self) -> ty.List[ty.Dict[str, orm.Float]]:
+        """
+        Create the inputs for the evaluation function.
+        If the work chain is not initialized, the appropriate number of
+        inputs to fill the convergence window are generated.
+        Otherwise, one more input will be generated
+        """
+        if not self.initialized:
+            # Do enough calculations to fill the initial convergence window
+            num_new_iters = self.convergence_window
+            self.initialized = True
+        else:
+            num_new_iters = self._num_new_iters
+
+        self.current_index += num_new_iters
+        inputs = [{
+            self.input_key: orm.Float(self.input_values[i])
+        } for i in range(self.current_index - num_new_iters, self.current_index)]
+
+        return inputs
+
+    def _update(self, outputs: ty.Dict[int, ty.Any]) -> None:
+        """
+        Update the state of the engine by saving all the output values and
+        storing them in the result_values property
+        """
+        output_values = outputs.values()  # Don't need keys (AiiDA UUIDs)
```

Comment:
The keys here are actually internal to `aiida-optimize` - they map inputs to outputs. We should sort by key, because that's the order in which the calculations were launched.

Maybe we're lucky with dictionaries remaining sorted by insertion order all the way through, but better safe than sorry.

Reply:
Yeah, sometimes I get complacent with the new versions of Python which preserve key order. I've added a sorting step for the keys, but please check my implementation.

Reply:
I mean from py3.7+ onwards that's fine, I was more worried about the YAML round-trip here. Doesn't hurt to sort anyway.

Reply:
Sorting looks good now :+1: